[doc] Updated documentations for implicit type casting rules (#4885)
taichi-dev/taichi: docs/lang/articles/basic/type.md

#### Implicit type casting in binary operations

> Taichi implements its own implicit type casting rules for binary operations, which is slightly different from that for C programming language: [implicit conversion rules](https://en.cppreference.com/w/c/language/conversion) and [binary operation](https://en.wikipedia.org/wiki/Binary_operation).

Reviewer: ```suggestion
Taichi implements its own implicit type casting rules for binary operations, which are slightly different from [those for the C programming language](https://en.cppreference.com/w/c/language/conversion).
```

Author: Done. Thanks!

In general we have three rules with descending priority:

1. integral OP floating_point -> floating_point
- `i32 + f32 -> f32`
- `i16 + f16 -> f16`

2. low_precision_bits OP high_precision_bits -> high_precision_bits
- `i16 + i32 -> i32`
- `u8 + u16 -> u16`

> 3. signed OP unsigned -> signed

Reviewer: The result should be unsigned..

Author: Many thanks!
